### PR TITLE
[RFC] ipfs for history sync

### PR DESCRIPTION
**Pull request by cmr**
See PR body.

### CODE REVIEW

**Comment by cmr** [comment]
the ipfs binary is 37MB.

**Comment by bkase** [comment]
I like this 👍

**Comment by jkrauska** [comment]
What does this do for mobile development? (how portable is this dependency?)

File: rfcs/0000-ipfs-net.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  val remove : handle:string -> unit Deferred.Or_error.t
+  (** Remove the data referred to by [handle] from local storage and stop offering it to the network. *)
+
+  val fetch : handle:string -> Bigstring.t Deferred.Or_error.t
+  (** Attempt to retrieve a piece of data. *)
+end
+```
+
+Internally, these are implemented using ipfs's `block` API. The three main methods correspond exactly to `block/put`, `block/rm`, and `block/get` respectively. We will spwan and manage our own ipfs daemon that isn't necessarily connected to the wider ipfs network (by not using the default bootstrap list). The provided blocks persist past restarts - the frontier will have to keep track of which block it stops caring about and explicitly `remove` them.
```
Comment (psteckler) [inline rfcs/0000-ipfs-net.md]: "spawn"

File: rfcs/0000-ipfs-net.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+ipfs internally uses blocks of size 2¹⁸ (256K). Our blocks are smaller than that, so this interface should be sufficient. For larger information, the higher level
```
Comment (psteckler) [inline rfcs/0000-ipfs-net.md]: seems like some text got deleted here

File: rfcs/0000-ipfs-net.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
+## Summary
+[summary]: #summary
+
+Use ipfs as a content sharing network for history sync. There is also space to use their "gossipsub" pubsub protocol (when it's finished) for efficient broadcasting.
```
Comment (psteckler) [inline rfcs/0000-ipfs-net.md]: Maybe add a link to IPFS docs somewhere.

File: rfcs/0000-ipfs-net.md (hunk 1 above)
Comment (johnwu93) [inline rfcs/0000-ipfs-net.md]: It would be helpful to discuss the technical details of a `block` for IPFS. Also, include a  link of the doc to `block`

**Comment by cmr** [comment]
@jkrauska wrote:

> What does this do for mobile development? (how portable is this dependency?)

I've addressed that a little with the most recent commit. ipfs runs on phones OK, but we'd want to turn off a bunch of features since we only need to `get`.

File: rfcs/0000-ipfs-net.md (hunk 1 above)
Comment (cmr) [reply rfcs/0000-ipfs-net.md]: Does the new text satisfy you? It's not much, but blocks aren't very complicated. Any more details would require diving pretty deep into ipfs minutiae.

**Comment by cmr** [comment]
> This network seems like it would have similar security requirements to our underlying peer discovery one. I recall the S-Kademlia style cryptographic node IDs being called out in the cardano documentation (see pdfs.semanticscholar.org/3165/2823ca71520038773346b6e5bbfadc5c8419.pdf ). It's important for it to be hard to choose these node ids to protect against certain attacks.

We don't have a choice of messing that up with the ipfs interface, it's already modified S/Kademlia. This still leaves the question of whether our identity keypairs need to be long lived.

**Comment by cmr** [comment]
Our data chunks are not smaller than `2**18` bytes (I think we put the bound around 2MB?). Therefore we cannot use the raw `block` API, and will need to use the higher-level file APIs. I still need to look into how that will work.

**Comment by cmr** [comment]
Also, an ipfs node is also a lot harder to bundle into an iOS app than I initially thought.

Their JS DHT now is interoperable with their Go one.

**Pull request closed without merging**